CQ-4299472 Add UI Testing submodule

catalan-adobe opened the pull request:
## Description

* Add `ui.tests` submodule to the archetype
* `ui.tests` is a test module relying on Node.Js / Webdriver.IO (testing framework).
* Contains sample basic UI tests (also a example of upload test)
* Contains configuration to package and execute the tests on Docker (useful both for local testing and for testing in AEMaaCS)
* The submodule project is also prepared to be used in AEMaaCS automation (when feature will be available in CloudManager)

## Motivation and Context

CQ-4299472 [Cust UI] Archetype should create customer ui tests module

As a new customer, I would like to have the aem-project-archetype generate a UI Tests submodule to help me start writing UI tests for AEM.

Acceptance Criteria:
* aem-project-archetype will generate a Customer ui tests maven submodule with wdio
* Minimal generic sample tests in the module

## How Has This Been Tested?

1. SNAPSHOT version of the archetype project has been built and installed locally.
2. Created a new project based on the new snapshot to confirm new module is created as expected
3. Tested the new module (ui.tests): build + execution of the tests

## Screenshots (if appropriate):

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [X] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [X] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [X] I have updated the documentation accordingly.
- [X] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [X] All new and existing tests passed.

catalan-adobe commented:
@bpauli, thanks for the review!
About the dependency on AEMaaCS. It's already not dependent to it. This project will perfectly work with a local CQ quickstart instance.
The Docker part is indeed required for AEMaaCS but I would also suggest to keep it for non-cloud:
* The Maven build/execute profiles (relying on the Dockerfile and the Docker compose files) are I think a nice feature for user. It's an easy way for them to quickly build and execute everything locally.
* Advanced users can still look at the `test-module`. There they will find npm scripts and wdio configurations for local execution (relying on `selenium-standalone` wdio service to provide Selenium endpoints).

About the extra test on the test page, very good idea, I'll look to add one.

catalan-adobe commented on src/main/archetype/ui.tests/test-module/specs/aem/sites.js:
```diff
@@ -0,0 +1,60 @@
+
+
+const AEM_SITES_PATH = '/sites.html';
+const AEM_SAMPLE_PAGE_PARENT = '/content/${appId}/us';
```
@bpauli, is it possible to get this file parsing the actual value of `appId`?

catalan-adobe replied:
Forget this. I tested it and variable is already being replaced in created project.

catalan-adobe commented:
@bpauli, quick summary on remaining items for this PR:
* Question about these tests being mainly for cloud case. As commented above, the tests are not only for cloud case. The Dockerfile is required for cloud case but they can be used for local case (targeting Quickstart jar).
Though, if it helps avoiding confusion, we could only add module `ui.tests` if Archetype is `cloud`.
* Added a test for the sample page created in `ui.content` but I would need a way to get the value of `appId` inside the test code (see comment above).
* Since your first review I did minor changes:
  * make them agnostic to AEM language
  * added linting feature with basic rules + fixed linting issues in code
Would you have time to look at these points?

dplaton commented:
@catalan-adobe I'm a bit confused about the purpose of the UI testing module in the archetype. The sample you added to the PR do some tests on the author instance, is this intended? Why would a customer need to test the authoring UI - consoles, etc. Shouldn't this test be part of the Quickstart already?

catalan-adobe commented:
@dplaton, good point! The idea of this module is to give a ready-to-use UI testing project (with building system, frameworks, ...) on which they "just" need to add their tests.
About the sample tests we provide: the great thing with author tests is that customers can get them executed on both generic AEM instances but also on their AEMaaCS deployments. These same tests are meant to work on any type of author instances (author experience is the same).
But for publish case, experience there depends on the customer application. We could write UI tests for the publish application from the archetype but customers could not get these tests executed on their application. Also, customer's main use case might not be about testing author but I can imagine they have special cases (in area of authoring and/or assets) where the provided tests could help.

We could add tests for publish use case but then we either:
* Create tests which target adobe.com so tests could be executed OOTB by customers (though it might not help confusion :))
* Create tests for publish case based on the archetype sample application but then experience of this module is not OOTB. We need to document that users needs to first install sample app. Also these tests will not work on AEMaaCS deployments.

msagolj commented on src/main/archetype/ui.tests/README.md:
```diff
@@ -0,0 +1,63 @@
+UI Test Project compatible with AEMaaCS
+=======================================
+
+[TODO]
```
Looks like the description is missing.

msagolj commented on src/main/archetype/ui.tests/README.md:
```diff
@@ -0,0 +1,63 @@
+
+[TODO]
+
+## Structure
```
In the generated project, all the titles (##) are gone. Looks like the README file is being interpreted and the headings removed.

msagolj commented on src/main/archetype/ui.tests/docker-compose-wdio-firefox.yaml:
```diff
@@ -0,0 +1,44 @@
+    volumes:
+      - ./reports:/tmp/reports
+      - 'shared:/tmp/shared'
+    image: 'com.adobe.cq.testing/ui.tests:latest'
```
The docker image name is hard coded, causing the run test step to fail.

catalan-adobe commented:
@msagolj, as discussed with you, I refactored the module (and relevant documentation) to integrate it better with the project generated by the archetype.
I also made classic "Maven/Frontend/Node" build the default use case and moved Docker build as alternative/advanced use case.

catalan-adobe commented:
@msagolj, I fixed the last 2 issues you are getting:
1. Handle the AEM Survey dialog that appears randomly
2. Fix `sites.js` test for Windows platform.
FYI, this time I also tested everything on a Windows10 VM machine.

catalan-adobe commented:
@msagolj, I fixed the last 3 issues you encountered (mainly due to first time experience on a fresh AEM instance):

* Handle the AEM Tutorial overlays/popups
* Increase default timeouts because of slow first-time rendering of AEM pages
* Disabled "save password" Chrome popup

drewrobinson commented:
Is it possible to include a linting solution in this?

catalan-adobe commented:
@drewrobinson, the WDIO project/Js code implements ESLint linting (check `test-module/package.json`, `lint` NPM script).
